tiledb/sm/buffer/buffer.cc

```diff
@@ -42,49 +41,120 @@ using namespace tiledb::common;
 namespace tiledb {
 namespace sm {
 
-/* ****************************** */
-/*             MACROS             */
-/* ****************************** */
+// =======================================================
+//      BufferBase
+// =======================================================
 
-#define MAX(a, b) ((a) > (b) ? (a) : (b))
+BufferBase::BufferBase()
+    : data_(nullptr)
+    , size_(0)
+    , offset_(0){};
 
-/* ****************************** */
-/*   CONSTRUCTORS & DESTRUCTORS   */
-/* ****************************** */
+BufferBase::BufferBase(void* data, const uint64_t size)
+    : data_(data)
+    , size_(size)
+    , offset_(0){};
 
-Buffer::Buffer()
-    : Buffer(nullptr, 0) {
-  owns_data_ = true;
+BufferBase::BufferBase(const void* data, const uint64_t size)
+    // const_cast is safe here because BufferBase methods do not modify storage
+    : data_(const_cast<void*>(data))
+    , size_(size)
+    , offset_(0){};
+
+uint64_t BufferBase::size() const {
+  return size_;
+}
+
+void* BufferBase::nonconst_data() const {
+  return data_;
+}
+
+const void* BufferBase::const_data() const {
+  // const_cast is safe here because BufferBase methods do not modify storage
+  return const_cast<const void*>(data_);
+}
+
+void* BufferBase::nonconst_unread_data() const {
+  if (data_ == nullptr) {
+    return nullptr;
+  }
+  // Cast to byte type because offset is measured in bytes
+  return static_cast<int8_t*>(data_) + offset_;
 }
 
-Buffer::Buffer(void* data, const uint64_t size) {
+const void* BufferBase::const_unread_data() const {
+  // const_cast is safe here because BufferBase methods do not modify storage
+  return const_cast<const void*>(nonconst_unread_data());
+}
+
+uint64_t BufferBase::offset() const {
+  return offset_;
+}
+
+void BufferBase::reset_offset() {
   offset_ = 0;
-  alloced_size_ = 0;
-  owns_data_ = false;
-  data_ = data;
-  size_ = size;
 }
 
-Buffer::Buffer(const Buffer& buff) {
-  alloced_size_ = buff.alloced_size_;
+void BufferBase::set_offset(const uint64_t offset) {
+  if (offset > size_) {
+    assert(offset <= size_);
+  }
+  offset_ = offset;
```

Comment:
It's a violation of precondition to set the offset outside the buffer size, so if anything, we ought to throw an exception here. Constraining the value would not be good, as it would be masking a fault that needs to be uncovered.

Reply:
Hence the thought of logging a warning, ATM nothing is reported, just setting an invalid offset.

Reply:
I changed it to throwing an exception. `assert` would for debugging, but it resolves to empty in release mode.